File: color.go
package appclipcode

import (
	"fmt"
	"strings"
)

// Color represents an RGB color.
type Color struct {
	R, G, B uint8
}

// ParseHexColor parses a 6-digit hex color string like "FF3B30".
func ParseHexColor(s string) (Color, error) {
	s = strings.TrimPrefix(s, "#")
	if len(s) != 6 {
		return Color{}, fmt.Errorf("color must be 6 hex digits, got %q", s)
	}
	var c Color
	_, err := fmt.Sscanf(s, "%02x%02x%02x", &c.R, &c.G, &c.B)
	if err != nil {
		return Color{}, fmt.Errorf("invalid hex color %q: %w", s, err)
	}
	return c, nil
}

// Hex returns the color as a lowercase 6-digit hex string with # prefix.
func (c Color) Hex() string {
	return fmt.Sprintf("#%02x%02x%02x", c.R, c.G, c.B)
}

// Palette holds the three colors used in an App Clip Code.
type Palette struct {
	Foreground Color // data-color=0 arcs
	Background Color // background circle
	Third      Color // data-color=1 arcs
}

// Template is a predefined color template.
type Template struct {
	Index      int
	Foreground Color
	Background Color
	Third      Color
}

// 9 base palette presets. Each has fg-on-bg (even index) and bg-on-fg (odd index) variants.
var basePalettes = [9]Palette{
	{Foreground: Color{0x00, 0x00, 0x00}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0x88, 0x88, 0x88}},
	{Foreground: Color{0x77, 0x77, 0x77}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0xAA, 0xAA, 0xAA}},
	{Foreground: Color{0xFF, 0x3B, 0x30}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0xFF, 0x99, 0x99}},
	{Foreground: Color{0xEE, 0x77, 0x33}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0xEE, 0xBB, 0x88}},
	{Foreground: Color{0x33, 0xAA, 0x22}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0x99, 0xDD, 0x99}},
	{Foreground: Color{0x00, 0xA6, 0xA1}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0x88, 0xDD, 0xCC}},
	{Foreground: Color{0x00, 0x7A, 0xFF}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0x77, 0xBB, 0xFF}},
	{Foreground: Color{0x58, 0x56, 0xD6}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0xBB, 0xBB, 0xEE}},
	{Foreground: Color{0xCC, 0x73, 0xE1}, Background: Color{0xFF, 0xFF, 0xFF}, Third: Color{0xEE, 0xBB, 0xEE}},
}

// Templates returns all 18 predefined color templates.
func Templates() []Template {
	templates := make([]Template, 18)
	for i, p := range basePalettes {
		// Even index: white foreground on colored background
		templates[i*2] = Template{
			Index:      i * 2,
			Foreground: Color{0xFF, 0xFF, 0xFF},
			Background: p.Foreground,
			Third:      p.Third,
		}
		// Odd index: colored foreground on white background
		templates[i*2+1] = Template{
			Index:      i*2 + 1,
			Foreground: p.Foreground,
			Background: p.Background,
			Third:      p.Third,
		}
	}
	return templates
}

// TemplateByIndex returns the palette for a given template index (0-17).
func TemplateByIndex(index int) (Palette, error) {
	templates := Templates()
	if index < 0 || index >= len(templates) {
		return Palette{}, fmt.Errorf("template index must be 0-17, got %d", index)
	}
	t := templates[index]
	return Palette{
		Foreground: t.Foreground,
		Background: t.Background,
		Third:      t.Third,
	}, nil
}

File: js/src/color.ts
export class Color {
  constructor(
    public readonly r: number,
    public readonly g: number,
    public readonly b: number,
    public readonly a: number = 0xff,
  ) {}

  hex(): string {
    if (this.a === 0xff) {
      return `#${toHex(this.r)}${toHex(this.g)}${toHex(this.b)}`;
    }

    return `#${toHex(this.r)}${toHex(this.g)}${toHex(this.b)}${toHex(this.a)}`;
  }
}

export interface Palette {
  foreground: Color;
  background: Color;
  third: Color;
}

export interface Template extends Palette {
  index: number;
}

function toHex(n: number): string {
  return n.toString(16).padStart(2, "0");
}

export function parseHexColor(value: string): Color {
  const normalized = String(value).replace(/^#/, "");
  if (!/^[0-9a-fA-F]{6}([0-9a-fA-F]{2})?$/.test(normalized)) {
    throw new Error(`color must be 6 or 8 hex digits, got ${JSON.stringify(normalized)}`);
  }

  return new Color(
    Number.parseInt(normalized.slice(0, 2), 16),
    Number.parseInt(normalized.slice(2, 4), 16),
    Number.parseInt(normalized.slice(4, 6), 16),
    normalized.length === 8 ? Number.parseInt(normalized.slice(6, 8), 16) : 0xff,
  );
}

export function midpointColor(a: Color, b: Color): Color {
  return new Color(
    Math.floor((a.r + b.r) / 2),
    Math.floor((a.g + b.g) / 2),
    Math.floor((a.b + b.b) / 2),
    Math.floor((a.a + b.a) / 2),
  );
}

const basePalettes: Palette[] = [
  { foreground: new Color(0x00, 0x00, 0x00), background: new Color(0xff, 0xff, 0xff), third: new Color(0x88, 0x88, 0x88) },
  { foreground: new Color(0x77, 0x77, 0x77), background: new Color(0xff, 0xff, 0xff), third: new Color(0xaa, 0xaa, 0xaa) },
  { foreground: new Color(0xff, 0x3b, 0x30), background: new Color(0xff, 0xff, 0xff), third: new Color(0xff, 0x99, 0x99) },
  { foreground: new Color(0xee, 0x77, 0x33), background: new Color(0xff, 0xff, 0xff), third: new Color(0xee, 0xbb, 0x88) },
  { foreground: new Color(0x33, 0xaa, 0x22), background: new Color(0xff, 0xff, 0xff), third: new Color(0x99, 0xdd, 0x99) },
  { foreground: new Color(0x00, 0xa6, 0xa1), background: new Color(0xff, 0xff, 0xff), third: new Color(0x88, 0xdd, 0xcc) },
  { foreground: new Color(0x00, 0x7a, 0xff), background: new Color(0xff, 0xff, 0xff), third: new Color(0x77, 0xbb, 0xff) },
  { foreground: new Color(0x58, 0x56, 0xd6), background: new Color(0xff, 0xff, 0xff), third: new Color(0xbb, 0xbb, 0xee) },
  { foreground: new Color(0xcc, 0x73, 0xe1), background: new Color(0xff, 0xff, 0xff), third: new Color(0xee, 0xbb, 0xee) },
];

export function templates(): Template[] {
  const result: Template[] = [];
  for (let i = 0; i < basePalettes.length; i += 1) {
    const palette = basePalettes[i];
    result.push({
      index: i * 2,
      foreground: new Color(0xff, 0xff, 0xff),
      background: palette.foreground,
      third: palette.third,
    });
    result.push({
      index: i * 2 + 1,
      foreground: palette.foreground,
      background: palette.background,
      third: palette.third,
    });
  }
  return result;
}

export function templateByIndex(index: number): Palette {
  const all = templates();
  if (index < 0 || index >= all.length) {
    throw new Error(`template index must be 0-17, got ${index}`);
  }
  const template = all[index];
  return {
    foreground: template.foreground,
    background: template.background,
    third: template.third,
  };
}

export function findThirdColor(foreground: Color, background: Color): Color {
  const alpha = Math.floor((foreground.a + background.a) / 2);
  for (const palette of basePalettes) {
    if (sameRGB(palette.foreground, foreground) && sameRGB(palette.background, background)) {
      return new Color(palette.third.r, palette.third.g, palette.third.b, alpha);
    }
    if (sameRGB(palette.foreground, background) && sameRGB(palette.background, foreground)) {
      return new Color(palette.third.r, palette.third.g, palette.third.b, alpha);
    }
  }
  return midpointColor(foreground, background);
}

function sameRGB(a: Color, b: Color): boolean {
  return a.r === b.r && a.g === b.g && a.b === b.b;
}
